### PR TITLE
/etc/hosts record not created automatically in AY installation

### PR DESCRIPTION
**Pull request by mchf**
https://bugzilla.suse.com/show_bug.cgi?id=1178260

## Problem ##

If you have static interface configuration and a hostname definition in AY profile it is expected to end up with record like

<ip> <hostname> ...

in /etc/hosts on the installed system

## Solution ##

Had two phases - for first stage only installation where a minor tweak in return value was enough. Second stage fix was a little bit more complicated as it was re-written to  use new network-ng. Also both fixes work only for installation workflows and not when running AY in installed system.

### CODE REVIEW

File: src/modules/Lan.rb [hunk 1]
```diff
@@ -643,10 +643,20 @@ def Import(settings)
...
       # Ensure that the /etc/hosts has been read to no blank out it in case of
       # not defined <host> section (bsc#1058396)
       Host.Read
 
+      # FIXME: hotfix for Host::ResolveHostnameToStaticIPs - the functionality is
+      # not implemented in new config readers/writers which are used in second stage
+      static_connections = config.connections.select { |c| c.static? }
+      static_ips = static_connections.map { |c| c.ip.address.address.to_s }
+
+      log.info("Lan::Import: updating /etc/hosts with static ips #{static_ips} and hostname #{config.hostname.static}")
```
Comment (mchf) [inline src/modules/Lan.rb:656]: this should go away ... it is just check whether they used the patch correctly ;-)

File: src/modules/Lan.rb [hunk 2]
```diff
@@ -643,10 +643,20 @@ def Import(settings)
       end
 
       NetworkConfig.Import(settings["config"] || {})
+
       # Ensure that the /etc/hosts has been read to no blank out it in case of
       # not defined <host> section (bsc#1058396)
       Host.Read
 
+      # FIXME: hotfix for Host::ResolveHostnameToStaticIPs - the functionality is
```
Comment (teclator) [inline src/modules/Lan.rb:651]: In general hhe relation between the Host and connections hostname has to be reevaluated, at least the method description suggested that the method for using the hostname associated to the static IPs was installation specific. 

At the end it is also expected in AutoYaST which makes sense when there is no `<hosts>` section

File: src/modules/Lan.rb [hunk 3]
```diff
@@ -643,10 +643,20 @@ def Import(settings)
...
+      # not implemented in new config readers/writers which are used in second stage
+      static_connections = config.connections.select { |c| c.static? }
```
Comment (teclator) [inline src/modules/Lan.rb:653]: I will put all this logic in a separate method which deserve a unit test.

**Comment by teclator** [comment]
In general LGTM, the different corner cases should be covered by the AutoYaST reader (merge of configs in case of keep installation and so on)

What could be problematic is when the list of hosts is given, in that case the user explicitly decided to not add the static IP alias and we are not overridden it, isn't it?

**Comment by mchf** [comment]
> What could be problematic is when the list of hosts is given, in that case the user explicitly decided to not add the static IP alias and we are not overridden it, isn't it?

well I hardly doubt that it has been ever specified. In my POV explicitly defined thing from the hosts section should win

File: src/lib/network/network_autoyast.rb [hunk 4]
```diff
@@ -157,6 +157,8 @@ def configure_hosts(write: false)
       end
       hosts_config = hosts_config.to_h.delete_if { |k, v| k.empty? || v.empty? }
 
+      return false if hosts_config.empty?
```
Comment (teclator) [inline src/lib/network/network_autoyast.rb:160]: With this fix, the behavior should be the same than in SLE-12-X that is, should fallback to the ResolveHostname... behavior, other fix than that affecting the second stage will be a real change of behavior.

**Comment by teclator** [comment]
> > What could be problematic is when the list of hosts is given, in that case the user explicitly decided to not add the static IP alias and we are not overridden it, isn't it?
> 
> well I hardly doubt that it has been ever specified. In my POV explicitly defined thing from the hosts section should win

It is not so, because we read the current file in order to not get rid of defaults, so, basically it preserve what is there, in case of <networking> and <host> section, the entries will be added by lan_auto and then not remove by hosts_auto

**Comment by teclator** [comment]
> > > What could be problematic is when the list of hosts is given, in that case the user explicitly decided to not add the static IP alias and we are not overridden it, isn't it?
> > 
> > 
> > well I hardly doubt that it has been ever specified. In my POV explicitly defined thing from the hosts section should win
> 
> It is not so, because we read the current file in order to not get rid of defaults, so, basically it preserve what is there, in case of and section, the entries will be added by lan_auto and then not remove by hosts_auto

After some discussion, apparently the behavior we are trying to introduce was already there ins SLE-11, but checking the code, I have not found references.

In SLE-11, there was a separate client for specifying the hostname, and the domain name, in case that the network configuration was omitted it added the entry to the /etc/hosts but only in a normal installation, never in AutoYaST

That was modified in https://github.com/yast/yast-network/pull/200, which added the call to configure_hosts, but was never part of lan_auto, that is, the behavior was never part of the `Second Stage`.

So, in SLE-12-X, the /etc/hosts entry added automatically based on the configured static IP addresses are `First Stage` specific. Thus, in the bug, the entry added is based on the linuxrc static IP address configuration.

if the AutoYaST installation is done with `ifcfg=*=dhcp`, no entry will be added to /etc/hosts although we specify a different option in the profile.

Of course, it would be nice to propose it in case that there is no hosts section during the second stage, but lan client is not aware of the host section.

That is my main concern about the behavior change. If it is decided that it is preferable to add the host entry always even if there is a declared host section which does not contain it then ok, if not would be preferable that the entry is declared explicitly and have a proper documentation in AutoYaST.

In SLE-15-SP3 it is different because we are doing everything in the first stage and we can check whether the hosts_section is defined or not to propose anything.

File: package/yast2-network.changes [hunk 5]
```diff
@@ -8,6 +16,8 @@ Fri Feb 26 15:23:33 UTC 2021 - Knut Anderssen <kanderssen@suse.com>
     exist.
 - 4.2.92
 
+-------------------------------------------------------------------
+>>>>>>> 6de30074... Updated changelog
```
Comment (teclator) [inline package/yast2-network.changes:20]: Leftover

Comment (mchf) [reply package/yast2-network.changes]: oh %&*!

File: src/lib/y2network/clients/auto.rb [hunk 6]
```diff
@@ -116,6 +118,23 @@ def merge_current_config?
         !!Yast::Lan.autoinst.keep_install_network
       end
 
+      # Whatever is needed in /etc/hosts during AY installation
+      def update_etc_hosts
+        config = Yast::Lan.find_config(:yast)
+        return if !config
+
+        static_connections = config.connections.select(&:static?)
+        static_ips = static_connections.map { |c| c.ip.address.address.to_s }
```
Comment (teclator) [inline src/lib/y2network/clients/auto.rb:127]: Why not assign the hostname to the connection? the writer should take care of updating the hosts entry, or is there some corner case? Isn't possible that otherwise the hostname is overwritten by the specific writer?

https://github.com/yast/yast-network/blob/SLE-15-SP2/src/lib/y2network/sysconfig/connection_config_writers/base.rb#L57

Comment (mchf) [reply src/lib/y2network/clients/auto.rb:127]: i don't know ... I'm not that familiar with connections yet. But will give it a try.

File: src/lib/y2network/clients/auto.rb [hunk 7]
```diff
@@ -89,6 +89,8 @@ def import(profile)
 
         Yast::Lan.Import(modified_profile)
 
+        update_etc_hosts
```
Comment (teclator) [inline src/lib/y2network/clients/auto.rb:92]: I thought we agreed on doing it only in case of a Second stage but not as a general approach.

Comment (mchf) [reply src/lib/y2network/clients/auto.rb:92]: yeah ... just checking whether you pay attention ... just joking ... I don't know where it disappeared, seems like a copy paste issue ... fixing it.

File: src/lib/y2network/clients/auto.rb [hunk 8]
```diff
@@ -116,6 +119,21 @@ def merge_current_config?
         !!Yast::Lan.autoinst.keep_install_network
       end
 
+      # Whatever is needed in /etc/hosts during AY installation
+      def update_etc_hosts
+        config = Yast::Lan.find_config(:yast)
+        return if !config
+
+        static_connections = config.connections.select(&:static?)
```
Comment (teclator) [inline src/lib/y2network/clients/auto.rb:127]: I would select the ones which are static and which the hostname is empty.

**Comment by teclator** [comment]
Although I would prefer to document it and add an specific entry in the hosts section in case of 2nd Stage, with the latest modifications I think that is ok.

**Review by teclator — APPROVED** [review]
LGTM
